### DISCUSSION TITLE
Expand or shrink shape to line

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#964] [Q&A]
Does anybody know how to expand or shrink vector shapes to a line?

I have seen that Apple Maps does this often with for example forest polygons. I looks like they are made larger or smaller to just nicely touch the road lines like for example in this curve:

![image](https://github.com/user-attachments/assets/1a19065f-753a-42e9-a1ed-66b3d20cdeed)

https://beta.maps.apple.com/?ll=47.43895624151364%2C8.006096755577289&spn=0.002407018447470932%2C0.00540248678916555

Or here, the settlement polygon under the buildings goes exactly to the road and not out into the field:

![image](https://github.com/user-attachments/assets/7d3456d1-2e9a-4346-95d1-1cafd45d9298)

https://beta.maps.apple.com/?ll=47.367397439296965%2C8.078616731881539&spn=0.0028911758158400858%2C0.006480356065736714

**Comment by wipfli** [comment]
I don't know if I really like this as a style element but I am curious how one could do it on a technical level...

**Reply by zstadler** [reply]
Are you asking about ways to modify the raw data or on is it about implementing this as part of the tile generation and its rendering?

**Reply by wipfli** [reply]
The latter.

**Comment by msbarry** [comment]
They might be using an encoding more like topojson where polygons and lines share arcs, which is much better for landcover polygons since you can simplify/smooth without introducing gaps and share between roads and lines...

You could potentially get something similar by splitting polygons by road lines and throwing out polygons that are too small/narrow. I don't think JTS has a built-in operation for this but you could get something similar by buffering the road lines by a small amount to turn them into polygons then subtracting them from landcover polgons, and throw out small leftover polygons.

**Reply by zstadler** [reply]
The idea @msbarry described handles well polygons that extend beyond a road by removing small parts "on the other side of the road"
For polygons that fail to touch a road you could add a buffer to the results and repeat.
